File: python/array-strings/345. Reverse Vowels of a String.py
"""
Given a string s, reverse only all the vowels in the string and return it.
The vowels are 'a', 'e', 'i', 'o', and 'u', and they can appear in both lower and upper cases, more than once.
Input: s = "IceCreAm"
Output: "AceCreIm"
The vowels in s are ['I', 'e', 'e', 'A']. On reversing the vowels, s becomes "AceCreIm".
"""


class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = ["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"]
        str = s.split()
        print(str)
        start = 0
        end = len(s) - 1
        while start < end:
            if s[start] in vowels and s[end] in vowels:
                s[start], s[end] = s[end], s[start]
            elif s[start] not in vowels:
                start += 1
            elif s[end] not in vowels:
                end -= 1

        return s


s = Solution()
print(s.reverseVowels("IceCreAm"))

File: javascript/array-strings/345. Reverse Vowels of a String.js
/*
Given a string s, reverse only all the vowels in the string and return it.
The vowels are 'a', 'e', 'i', 'o', and 'u', and they can appear in both lower and upper cases, more than once.
Input: s = "IceCreAm"
Output: "AceCreIm"
The vowels in s are ['I', 'e', 'e', 'A']. On reversing the vowels, s becomes "AceCreIm".
*/

/**
 * @param {string} s
 * @return {string}
 */
function reverseVowels(s) {
    let vowels = ["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"];
    let str = s.split('');
    console.log(str);
    let start = 0;
    let end = s.length - 1;
    
    while (start < end) {
        if (vowels.includes(s[start]) && vowels.includes(s[end])) {
            [str[start], str[end]] = [str[end], str[start]];
            start++;
            end--;
        } else if (!vowels.includes(s[start])) {
            start++;
        } else if (!vowels.includes(s[end])) {
            end--;
        }
    }

    return str.join('');
}

// Test case
console.log(reverseVowels("IceCreAm")); 